broadlink/climate.py

```diff
@@ -238,3 +238,324 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        
+        Args:
+            state (bool): power
+            target_temp (float): temperature set point 16<n<32
+            mode (str): cooling, heating, fan, dry, auto
+            speed (str): mute, low, mid, high, turbo (available only for cooling) 
+            swing_h (str): ON, OFF
+            swing_v (str): ON, OFF, 1, 2, 3, 4, 5 (fixed positions)
+            sleep (bool) 
+            display (bool)
+            health (bool)
+            cmnd_0d_rmask (int): unknown
+            cmnd_0e_rmask (int): unknown
+            cmnd_18 (int): unknown
+            checksum_lbit (int): subtracted from the left byte of the checksum
+        """
+        
+        PREFIX = [0x19, 0x00, 0xbb, 0x00, 0x06, 0x80, 0x00, 0x00, 0x0f, 0x00, 0x01, 0x01] # 12B
+        MIDDLE = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] # 13B + 2B checksum
+        SUFFIX = [0, 0, 0, 0, 0] # 5B
+        payload = PREFIX + MIDDLE + SUFFIX
+
+        assert ((target_temp >= 16) and (target_temp <= 32) and ((target_temp * 2) % 1 == 0))
+
+        if (swing_v == 'OFF'):
+            swing_L = 0b111
+        elif (swing_v == 'ON'):
+            swing_L = 0b000
+        else:
+            raise ValueError('unrecognized swing vertical value {}'.format(swing_v))
+
+        if (swing_h == 'OFF'):
+            swing_R = 0b111
+        elif (swing_h == 'ON'):
+            swing_R = 0b000
+        elif (swing_h >= 0 and swing_v <= 5):
+            swing_R = str(swing_v)
+        else:
+            raise ValueError('unrecognized swing horizontal value {}'.format(swing_h))
+
+        if (speed == 'low'):
+            speed_L, speed_R = 0x60, 0x00
+        elif (speed == 'mid'):
+            speed_L, speed_R = 0x40, 0x00
+        elif (speed  == 'high'):
+            speed_L, speed_R = 0x20, 0x00
+        elif (speed == 'mute'):
+            speed_L, speed_R = 0x40, 0x80
+            assert (mode == 'F')
+        elif (speed == 'turbo'):
+            speed_R = 0x40
+            speed_L = 0x20 # doesn't matter
+        elif (speed == 'auto'):
+            speed_L, speed_R = 0xa0, 0x00
+        else:
+            raise ValueError('unrecognized speed value: {}'.format(speed))
+
+        if (mode == 'auto'):
+            mode_1 = 0x00
+        elif (mode == 'cooling'):
+            mode_1 = 0x20
+        elif (mode == 'drying'):
+            mode_1 = 0x40
+            cmnd_0e_rmask = 0x16
+        elif (mode == 'heating'):
+            mode_1 = 0x80
+        elif (mode == 'fan'):
+            mode_1 = 0xc0
+            target_temp = 24.0
+            if speed == 'turbo':
+                raise ValueError('speed cannot be {} in fan mode'.format(speed))
+        else:
+            raise ValueError('unrecognized mode value: {}'.format(mode))
+
+        payload[0x0c] = ((int(target_temp) - 8 << 3) | swing_L)
+        payload[0x0d] = (int(swing_R) << 5) | cmnd_0d_rmask
+        payload[0x0e] = (0b10000000 if (target_temp % 1 == 0.5) else 0b0) | cmnd_0e_rmask
```

Comment:
Let me see if I understand. Is this bit activated when the target temperature is a multiple of 0.5?

Reply:
Nope, that's ending in .5

Reply:
Thank you, I got it now. But still look like a magic behavior. Why this? I wish I could understand what is going on behind these bits.

Reply:
Ah, now I see the full picture. You can only set target temperature with a 0.5 precision, correct? This bit indicates the presence of this decimal value.

Reply:
> This bit indicates the presence of this decimal value.

Exactly.

Reply:
We need to round the target temp before doing this to make this method more robust:
```python3
target_temp = round(target_temp * 2) / 2
```
If we don't round, 40.8 will set the temperature to 40, which is not the expected behavior.

Reply:
Rounding is nice! haven't thought of that. Replaced the assert with a descriptive ValueError.